src/ui/public/state_management/__tests__/monitor_changes.js

```diff
@@ -0,0 +1,107 @@
+import expect from 'expect.js';
+import sinon from 'sinon';
+import monitorStateChanges from 'ui/state_management/monitor_changes';
+import SimpleEmitter from 'ui/utils/SimpleEmitter';
+
+describe('monitorStateChanges', function () {
+  const noop = () => {};
+  const eventTypes = [
+    'save_with_changes',
+    'reset_with_changes',
+    'fetch_with_changes',
+  ];
+  let mockState;
+
+  beforeEach(() => {
+    mockState = new SimpleEmitter();
+  });
+
+  describe('sanity checks', function () {
+    it('should throw if not given a handler function', function () {
+      const fn = () => monitorStateChanges(mockState, 'not a function');
+      expect(fn).to.throwException(/must be a listener function/);
+    });
+
+    it('should throw if not given a valid cleanup function', function () {
+      const fn = () => monitorStateChanges(mockState, noop, 'not a function');
+      expect(fn).to.throwException(/must be a cleanup function/);
+    });
+  });
+
+  describe('calling the event handler', function () {
+    eventTypes.forEach((eventType) => {
+      describe(eventType, function () {
+        let handlerFn;
+
+        beforeEach(() => {
+          handlerFn = sinon.stub();
+        });
+
+        it('should get called', function () {
+          monitorStateChanges(mockState, handlerFn, noop);
+          sinon.assert.notCalled(handlerFn);
+          mockState.emit(eventType);
+          sinon.assert.calledOnce(handlerFn);
+        });
+
+        it('should be passed the event type', function () {
+          monitorStateChanges(mockState, handlerFn, noop);
+          mockState.emit(eventType);
+          const args = handlerFn.firstCall.args;
+          expect(args[0]).to.equal(eventType);
+        });
+
+        it('should be passed the changed keys', function () {
+          const keys = ['one', 'two', 'three'];
+          monitorStateChanges(mockState, handlerFn, noop);
+          mockState.emit(eventType, keys);
+          const args = handlerFn.firstCall.args;
+          expect(args[1]).to.equal(keys);
+        });
+      });
+    });
+  });
+
+  describe('cleanup handler', function () {
+    let cleanFn;
+
+    beforeEach(() => {
+      cleanFn = sinon.stub();
+    });
+
+    describe('handler function', function () {
```

Comment:
I think we could remove this describe block since it doesn't add additional information.